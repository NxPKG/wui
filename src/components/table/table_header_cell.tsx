/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, {
  FunctionComponent,
  HTMLAttributes,
  ThHTMLAttributes,
} from 'react';
import classNames from 'classnames';

import { WuiScreenReaderOnly } from '../accessibility';
import { CommonProps, NoArgCallback } from '../common';
import { WuiIcon } from '../icon';
import { resolveWidthAsStyle } from './utils';
import { WuiInnerText } from '../inner_text';

import {
  HorizontalAlignment,
  LEFT_ALIGNMENT,
  RIGHT_ALIGNMENT,
  CENTER_ALIGNMENT,
} from '../../services';
import { WuiI18n } from '../i18n';

export type TableHeaderCellScope = 'col' | 'row' | 'colgroup' | 'rowgroup';

type Props = CommonProps &
  Omit<ThHTMLAttributes<HTMLTableHeaderCellElement>, 'align' | 'scope'> & {
    align?: HorizontalAlignment;
    /**
     * Set `allowNeutralSort` on WuiInMemoryTable to false to force column
     * sorting.  WuiBasicTable always forces column sorting.
     */
    allowNeutralSort?: boolean;
    /**
     * _DEPRECATED: use `mobileOptions.show = false`_ Indicates if the
     * column should not show for mobile users (typically hidden because a
     * custom mobile header utilizes the column's contents)
     */
    hideForMobile?: boolean;
    /**
     * _DEPRECATED: use `mobileOptions.only = true`_ Indicates if the
     * column was created to be the row's heading in mobile view (this
     * column will be hidden at larger screens)
     */
    isMobileHeader?: boolean;
    isSortAscending?: boolean;
    isSorted?: boolean;
    /**
     * Mobile options for displaying differently at small screens
     */
    mobileOptions?: {
      /**
       * If false, will not render the column at all for mobile
       */
      show?: boolean;
      /**
       * Only show for mobile? If true, will not render the column at all
       * for desktop
       */
      only?: boolean;
    };
    onSort?: NoArgCallback<void>;
    scope?: TableHeaderCellScope;
    width?: string | number;
  };

export const WuiTableHeaderCell: FunctionComponent<Props> = ({
  children,
  align = LEFT_ALIGNMENT,
  onSort,
  isSorted,
  isSortAscending,
  allowNeutralSort,
  className,
  scope = 'col',
  mobileOptions = {
    show: true,
  },
  width,
  // Soon to be deprecated for {...mobileOptions}
  isMobileHeader,
  hideForMobile,
  style,
  ...rest
}) => {
  const classes = classNames('wuiTableHeaderCell', className, {
    'wuiTableHeaderCell--hideForDesktop': mobileOptions.only || isMobileHeader,
    'wuiTableHeaderCell--hideForMobile': !mobileOptions.show || hideForMobile,
  });

  const contentClasses = classNames('wuiTableCellContent', className, {
    'wuiTableCellContent--alignRight': align === RIGHT_ALIGNMENT,
    'wuiTableCellContent--alignCenter': align === CENTER_ALIGNMENT,
  });

  const styleObj = resolveWidthAsStyle(style, width);

  const CellComponent = children ? 'th' : 'td';

  if (onSort) {
    const buttonClasses = classNames('wuiTableHeaderButton', {
      'wuiTableHeaderButton-isSorted': isSorted,
    });

    let ariaSortValue: HTMLAttributes<any>['aria-sort'] = 'none';
    if (isSorted) {
      ariaSortValue = isSortAscending ? 'ascending' : 'descending';
    }

    function getScreenCasterDirection() {
      if (ariaSortValue === 'ascending') {
        return (
          <WuiI18n
            token="wuiTableHeaderCell.clickForDescending"
            default="Click to sort in descending order"
          />
        );
      }

      if (allowNeutralSort && ariaSortValue === 'descending') {
        return (
          <WuiI18n
            token="wuiTableHeaderCell.clickForUnsort"
            default="Click to unsort"
          />
        );
      }

      return (
        <WuiI18n
          token="wuiTableHeaderCell.clickForAscending"
          default="Click to sort in ascending order"
        />
      );
    }

    return (
      <CellComponent
        className={classes}
        scope={scope}
        role="columnheader"
        aria-sort={ariaSortValue}
        aria-live="polite"
        style={styleObj}
        {...rest}>
        <button
          type="button"
          className={buttonClasses}
          onClick={onSort}
          data-test-subj="tableHeaderSortButton">
          <span className={contentClasses}>
            <WuiInnerText>
              {(ref, innerText) => (
                <WuiI18n
                  token="wuiTableHeaderCell.titleTextWithSort"
                  default="{innerText}; Sorted in {ariaSortValue} order"
                  values={{ innerText, ariaSortValue }}>
                  {(titleTextWithSort: string) => (
                    <span
                      title={isSorted ? titleTextWithSort : innerText}
                      ref={ref}
                      className="wuiTableCellContent__text">
                      {children}
                    </span>
                  )}
                </WuiI18n>
              )}
            </WuiInnerText>

            {isSorted && (
              <WuiI18n
                token="wuiTableHeaderCell.sortedAriaLabel"
                default="Sorted in {ariaSortValue} order"
                values={{ ariaSortValue }}>
                {(sortedAriaLabel: string) => (
                  <WuiIcon
                    className="wuiTableSortIcon"
                    type={isSortAscending ? 'sortUp' : 'sortDown'}
                    size="m"
                    aria-label={sortedAriaLabel}
                  />
                )}
              </WuiI18n>
            )}
            <WuiScreenReaderOnly>
              <span>{getScreenCasterDirection()}</span>
            </WuiScreenReaderOnly>
          </span>
        </button>
      </CellComponent>
    );
  }

  return (
    <CellComponent
      className={classes}
      scope={scope}
      role="columnheader"
      style={styleObj}
      {...rest}>
      <div className={contentClasses}>
        <WuiInnerText>
          {(ref, innerText) => (
            <span
              title={innerText}
              ref={ref}
              className="wuiTableCellContent__text">
              {children}
            </span>
          )}
        </WuiInnerText>
      </div>
    </CellComponent>
  );
};
