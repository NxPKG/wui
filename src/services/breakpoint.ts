/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { keysOf } from '../components/common';

export type WuiBreakpointSize = 'xs' | 's' | 'm' | 'l' | 'xl';

export type WuiBreakpoints = {
  /**
   * Set the minimum window width at which to start to the breakpoint
   */
  [key in WuiBreakpointSize]: number;
};

export const BREAKPOINTS: WuiBreakpoints = {
  xl: 1200,
  l: 992,
  m: 768,
  s: 575,
  xs: 0,
};

export const BREAKPOINT_KEYS = keysOf(BREAKPOINTS);

/**
 * Given the current `width` and an object of `WuiBreakpoints`,
 * this function returns the string that is the name of the breakpoint key
 * that is less than or equal to the width
 *
 * @param {number} width Can either be the full window width or any width
 * @param {WuiBreakpoints} breakpoints An object with keys for sizing and values for minimum width
 * @returns {string | undefined} Name of the breakpoint key or `undefined` if a key doesn't exist
 */
export function getBreakpoint(
  width: number,
  breakpoints: WuiBreakpoints = BREAKPOINTS
): WuiBreakpointSize | undefined {
  // Find the breakpoint (key) whose value is <= windowWidth starting with largest first
  return keysOf(BREAKPOINTS).find(key => breakpoints[key] <= width);
}

/**
 * Given the current `width` and a max breakpoint key,
 * this function returns true or false if the `width` falls within the max
 * breakpoint or any breakpoints below
 *
 * @param {number} width Can either be the full window width or any width
 * @param {WuiBreakpointSize | number} max The named breakpoint or custom number to check against
 * @param {WuiBreakpoints} breakpoints An object with keys for sizing and values for minimum width
 * @returns {boolean} Will return `false` if it can't find a value for the `max` breakpoint
 */
export function isWithinMaxBreakpoint(
  width: number,
  max: WuiBreakpointSize | number,
  breakpoints: WuiBreakpoints = BREAKPOINTS
): boolean {
  if (typeof max === 'number') {
    return width <= max;
  } else {
    const currentBreakpoint = getBreakpoint(width, breakpoints);
    return currentBreakpoint
      ? breakpoints[currentBreakpoint] <= breakpoints[max]
      : false;
  }
}

/**
 * Given the current `width` and an array of breakpoint keys,
 * this function returns true or false if the `width` falls within
 * any of the named breakpoints
 *
 * @param {number} width Can either be the full window width or any width
 * @param {WuiBreakpointSize[]} sizes An array of named breakpoints
 * @param {WuiBreakpoints} breakpoints An object with keys for sizing and values for minimum width
 * @returns {boolean} Returns `true` if current breakpoint name is included in `sizes`
 */
export function isWithinBreakpoints(
  width: number,
  sizes: WuiBreakpointSize[],
  breakpoints: WuiBreakpoints = BREAKPOINTS
): boolean {
  const currentBreakpoint = getBreakpoint(width, breakpoints);
  return currentBreakpoint ? sizes.includes(currentBreakpoint) : false;
}
