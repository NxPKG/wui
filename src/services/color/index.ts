/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

export { isColorDark } from './is_color_dark';
export { isValidHex } from './is_valid_hex';
export { hexToHsv } from './hex_to_hsv';
export { hexToRgb } from './hex_to_rgb';
export { hsvToHex } from './hsv_to_hex';
export { hsvToRgb } from './hsv_to_rgb';
export { rgbToHex } from './rgb_to_hex';
export { rgbToHsv } from './rgb_to_hsv';
export {
  calculateContrast,
  calculateLuminance,
} from './luminance_and_contrast';
export {
  VISUALIZATION_COLORS,
  DEFAULT_VISUALIZATION_COLOR,
} from './visualization_colors';
export { colorPalette } from './color_palette';
export {
  wuiPaletteForLightBackground,
  wuiPaletteForDarkBackground,
  wuiPaletteColorBlind,
  wuiPaletteColorBlindBehindText,
  wuiPaletteForStatus,
  wuiPaletteForTemperature,
  wuiPaletteComplimentary,
  wuiPaletteNegative,
  wuiPalettePositive,
  wuiPaletteCool,
  wuiPaletteWarm,
  wuiPaletteGray,
} from './wui_palettes';
export { rgbDef, HSV, RGB } from './color_types';
