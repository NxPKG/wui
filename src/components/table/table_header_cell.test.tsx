/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../test/required_props';

import { WuiTableHeaderCell } from './table_header_cell';

import { RIGHT_ALIGNMENT, CENTER_ALIGNMENT } from '../../services';
import { WARNING_MESSAGE } from './utils';

test('renders WuiTableHeaderCell', () => {
  const component = (
    <WuiTableHeaderCell {...requiredProps}>children</WuiTableHeaderCell>
  );

  expect(render(component)).toMatchSnapshot();
});

test('renders td when children is null/undefined', () => {
  const component = render(<WuiTableHeaderCell {...requiredProps} />);

  expect(component).toMatchSnapshot();
});

describe('align', () => {
  test('defaults to left', () => {
    const component = <WuiTableHeaderCell />;

    expect(render(component)).toMatchSnapshot();
  });

  test('renders right when specified', () => {
    const component = <WuiTableHeaderCell align={RIGHT_ALIGNMENT} />;

    expect(render(component)).toMatchSnapshot();
  });

  test('renders center when specified', () => {
    const component = <WuiTableHeaderCell align={CENTER_ALIGNMENT} />;

    expect(render(component)).toMatchSnapshot();
  });
});

describe('width and style', () => {
  const _consoleWarn = console.warn;
  beforeAll(() => {
    console.warn = (...args: [any?, ...any[]]) => {
      // Suppress an expected warning
      if (args.length === 1 && args[0] === WARNING_MESSAGE) return;
      _consoleWarn.apply(console, args);
    };
  });
  afterAll(() => {
    console.warn = _consoleWarn;
  });

  test('accepts style attribute', () => {
    const component = (
      <WuiTableHeaderCell style={{ width: '20%' }}>Test</WuiTableHeaderCell>
    );

    expect(render(component)).toMatchSnapshot();
  });

  test('accepts width attribute', () => {
    const component = <WuiTableHeaderCell width="10%">Test</WuiTableHeaderCell>;

    expect(render(component)).toMatchSnapshot();
  });

  test('accepts width attribute as number', () => {
    const component = <WuiTableHeaderCell width={100}>Test</WuiTableHeaderCell>;

    expect(render(component)).toMatchSnapshot();
  });

  test('resolves style and width attribute', () => {
    const component = (
      <WuiTableHeaderCell width="10%" style={{ width: '20%' }}>
        Test
      </WuiTableHeaderCell>
    );

    expect(render(component)).toMatchSnapshot();
  });
});
