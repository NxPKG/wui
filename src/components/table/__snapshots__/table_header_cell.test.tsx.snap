// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`align defaults to left 1`] = `
<td
  class="wuiTableHeaderCell"
  role="columnheader"
  scope="col"
>
  <div
    class="wuiTableCellContent"
  >
    <span
      class="wuiTableCellContent__text"
    />
  </div>
</td>
`;

exports[`align renders center when specified 1`] = `
<td
  class="wuiTableHeaderCell"
  role="columnheader"
  scope="col"
>
  <div
    class="wuiTableCellContent wuiTableCellContent--alignCenter"
  >
    <span
      class="wuiTableCellContent__text"
    />
  </div>
</td>
`;

exports[`align renders right when specified 1`] = `
<td
  class="wuiTableHeaderCell"
  role="columnheader"
  scope="col"
>
  <div
    class="wuiTableCellContent wuiTableCellContent--alignRight"
  >
    <span
      class="wuiTableCellContent__text"
    />
  </div>
</td>
`;

exports[`renders WuiTableHeaderCell 1`] = `
<th
  aria-label="aria-label"
  class="wuiTableHeaderCell testClass1 testClass2"
  data-test-subj="test subject string"
  role="columnheader"
  scope="col"
>
  <div
    class="wuiTableCellContent testClass1 testClass2"
  >
    <span
      class="wuiTableCellContent__text"
    >
      children
    </span>
  </div>
</th>
`;

exports[`renders td when children is null/undefined 1`] = `
<td
  aria-label="aria-label"
  class="wuiTableHeaderCell testClass1 testClass2"
  data-test-subj="test subject string"
  role="columnheader"
  scope="col"
>
  <div
    class="wuiTableCellContent testClass1 testClass2"
  >
    <span
      class="wuiTableCellContent__text"
    />
  </div>
</td>
`;

exports[`width and style accepts style attribute 1`] = `
<th
  class="wuiTableHeaderCell"
  role="columnheader"
  scope="col"
  style="width:20%"
>
  <div
    class="wuiTableCellContent"
  >
    <span
      class="wuiTableCellContent__text"
    >
      Test
    </span>
  </div>
</th>
`;

exports[`width and style accepts width attribute 1`] = `
<th
  class="wuiTableHeaderCell"
  role="columnheader"
  scope="col"
  style="width:10%"
>
  <div
    class="wuiTableCellContent"
  >
    <span
      class="wuiTableCellContent__text"
    >
      Test
    </span>
  </div>
</th>
`;

exports[`width and style accepts width attribute as number 1`] = `
<th
  class="wuiTableHeaderCell"
  role="columnheader"
  scope="col"
  style="width:100px"
>
  <div
    class="wuiTableCellContent"
  >
    <span
      class="wuiTableCellContent__text"
    >
      Test
    </span>
  </div>
</th>
`;

exports[`width and style resolves style and width attribute 1`] = `
<th
  class="wuiTableHeaderCell"
  role="columnheader"
  scope="col"
  style="width:10%"
>
  <div
    class="wuiTableCellContent"
  >
    <span
      class="wuiTableCellContent__text"
    >
      Test
    </span>
  </div>
</th>
`;
